Src/Support/Google.Apis.Auth/OAuth2/TokenRefreshManager.cs

```diff
@@ -0,0 +1,176 @@
+                {
+                    _token = value;
+                }
+            }
+        }
+
+        internal async Task<string> GetAccessTokenForRequestAsync(CancellationToken cancellationToken)
+        {
+            Task<TokenResponse> refreshTask;
+            lock (_lock)
+            {
+                // If current token is not expired, then return it.
+                if (_token != null && !_token.IsExpired(_clock))
+                {
+                    return _token.AccessToken;
+                }
+                // Token refresh required, so start a task if not already started
+                if (_refreshTask == null)
+                {
+                    // Task.Run is required if the refresh completes synchonously,
+                    // otherwise _refreshTask is updated in an incorrect order.
+                    // And Task.Run also means it can be run here in the lock.
+                    _refreshTask = Task.Run(RefreshTokenAsync);
+                }
+                // If current token is not hard-expired, then return it.
+                if (_token != null && !_token.IsEffectivelyExpired(_clock))
+                {
+                    return _token.AccessToken;
+                }
+                refreshTask = _refreshTask;
+            }
+            // Otherwise block on refresh task.
+            if (cancellationToken.CanBeCanceled)
+            {
+                // Reasonably simple way of creating a task that can be cancelled, based on another task.
+                // (It would be nice if this were simpler.)
+                refreshTask = refreshTask.ContinueWith(
+                    task => ResultWithUnwrappedExceptions(task),
+                    cancellationToken,
+                    TaskContinuationOptions.None,
+                    TaskScheduler.Default);
+            }
+            return (await refreshTask.ConfigureAwait(false)).AccessToken;
+        }
+
+        internal static readonly TimeSpan[] RefreshTimeouts = new[] { TimeSpan.FromSeconds(12), TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(5) };
+        private async Task<TokenResponse> RefreshTokenAsync()
+        {
+            _logger.Debug("Token has expired, trying to get a new one.");
+            try
+            {
+                foreach (var timeout in RefreshTimeouts)
+                {
+                    var token = new CancellationTokenSource(timeout).Token;
+                    try
+                    {
+                        var success = await _refreshAction(token).ConfigureAwait(false);
+                        if (success)
+                        {
+                            _logger.Info("New access token was received successfully");
+                            return Token;
+                        }
+                        else
+                        {
+                            // If unsuccessful, but didn't timeout, then finish with failure.
+                            // TODO: Should this retry?
```

Comment:
I think it's probably worth it retrying, yes. It could easily be a transient server failure. We should make sure that the refresh action doesn't *itself* retry, mind you.

Reply:
OK, added retry here.
And verified that the none of the refresh actions themselves retry :)